Add more tests for loader (resolves LOAD-30, LOAD-23)

jessebrennan opened the pull request:
A lot has been added. Here's a summary of all of the changes in no particular order:
- If a file has a bad url the bundle will fail to load. There is a test to confirm this behavior.
- There is now a test to make sure that if a file is already uploaded it will not be loaded twice. (this resolves LOAD-23)
- Named tuples were prettied up 
- Minor clean up, out dated todos removed, etc
- If a timestamp has no zone info assume utc. There is also a test to confirm this behavior.
- Make the global arrays of bundles into instance variables. This is a little tidier than before.
- A test for _parse_all_bundles is added.
- Tests for other parsing functionality are added
- We now also test the loading mechanism directly (instead of just through integration tests.
- This includes testing a 'minimal' bundle.
- debug is turned off for testing

┆Issue is synchronized with this [JIRA Task](https://ucsc-cgl.atlassian.net/browse/LOAD-36)
┆Project Name: CGP-Data Loader
┆Issue Number: LOAD-36

jessebrennan committed:
Add more loader unit tests and fix bugs they caught

These tests were designed to catch parse errors with the loader. and
help ensure that everything needed from the bundle is specifically tested
against. It does a pretty good job, but there may be another place or two
where dependencies are not made explicit and tested against.

jessebrennan committed:
Separate out parsing code and test it better

Now the parsing is done in isolation to the loading. This will make more sense when making the upload multithreaded.

jessebrennan committed:
Turn off debug for tests

jessebrennan committed:
Add more tests for loader (resolved LOAD-30, LOAD-23)

A lot has been added. Here's a summary of all of the changes:
- If a file has a bad url the bundle will fail to load. There is a test 
  to confirm this behavior.
- There is now a test to make sure that if a file is already uploaded
  it will not be loaded twice.
- Named tuples were prettied up 
- Minor clean up, out dated todos removed, etc
- If a timestamp has no zone info assume utc. There is
  also a test to confirm this behavior.
- Make the global arrays of bundles into instance variables. This is a
  little tidier than before.
- A test for _parse_all_bundles is added.
- We now also test the loading mechanism directly (instead of just 
  through integration tests.
- This includes testing a 'minimal' bundle.

wshands reviewed (commented):
Looks pretty good to me

mikebaumann commented on loader/base_loader.py:
```diff
@@ -97,7 +101,7 @@ def upload_cloud_file_by_reference(self,
         e.g. "dg.4503/887388d7-a974-4259-86af-f5305172363d"
         :param file_version: a RFC3339 compliant datetime string
-        :return: file_uuid: str, file_version: str, filename: str
+        :return: file_uuid: str, file_version: str, filename: str, already_present
```
Please add type for `already_present`

mikebaumann commented on loader/base_loader.py:
```diff
@@ -368,6 +374,12 @@ def _upload_tagged_cloud_file_to_dss_by_copy(self, source_bucket: str,
         file_version = response.json().get('version', "blank")
 
+        # from dss swagger docs:
+        #   > 200 Returned when the file is already present and is identical to the file being uploaded.
```
I think the `>` should be removed from the comment, correct?
200 is returned when the file is already present and identical ...

Also, what will happen in the case of errors (e.g. 4xx or 5xx) return codes?

jessebrennan replied:
whoops! don't know what I was thinking here. 4xx and 5xx will cause the bundle to fail. If you think some other behavior should be desired let me know

mikebaumann commented on loader/base_loader.py:
```diff
@@ -377,6 +389,7 @@ def _upload_tagged_cloud_file_to_dss_by_copy(self, source_bucket: str,
 
             timeout = time.time() + timeout_seconds
             wait = 1.0
+            # FIXME busy wait could hopefully be replaced with asyncio
```
Can asyncio be used here, given this is taking place "behind" the DSS REST API, which is performing a cloud-to-cloud copy?

jessebrennan replied:
That's exactly *why* it can be used. any time there is a local wait in python another async func could be running. I'm not sure how often this code path is even used though. That research needs to be done to decide if adding asyncio here would actually make any substantial difference.

mikebaumann commented on loader/base_loader.py:
```diff
@@ -368,6 +374,12 @@ def _upload_tagged_cloud_file_to_dss_by_copy(self, source_bucket: str,
+            already_present = True
+        else:
+            already_present = False
         if response.status_code in (requests.codes.ok, requests.codes.created):
```
I think `requests.codes.ok` doesn't need to be in this set, correct?
If we get an ok (200) that means the file is already present and identical, so no sync copy will be performed, so we don't need to log that and wait for the `head_file` to indicate that the file appeared.

jessebrennan replied:
It does! I actually made this mistake originally thinking just as you did.  If you take it out then you get stuck in the else block bellow and time out. But you did make me realize this could be shortened to
```
already_present = response.status_code == requests.codes.ok
```

mikebaumann replied:
It doesn't! 😃 ... but additional code changes are needed in this case, which is what I had in mind but didn't explicitly describe.

There four specific return codes we want to handle here, each with a different log message:
- requests.codes.ok -> file already exists
- requests.codes.created -> Sync copy
- requests.codes.accepted -> Async copy 
- requests.codes.conflict -> File with same UUID+version exists but the new file is different from the existing file, so the new file is not loaded.

Let's talk about this before you start making changes to ensure the cases are clear.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -13,58 +11,125 @@
+        if version is None:
+            version = parse_version_key(file_info, 'created')
+        if version is None:
+            raise ParseError('Either bundle had no updated / created time or it was not rfc3339 compliant')
```
It seems like knowing which of these cases occurred (not present or invalid format) would be useful.
However, for current purposes, I think addressing this is optional.

jessebrennan replied:
a warning is logged which I think makes this clear in practice.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -13,58 +11,125 @@
+            raise ParseError(f'URL field not present in file_info: \n{file_info}')
+        urls = file_info['urls']
+        if len(urls) < 1:
+            # FIXME: How many cloud URLs do we ACTUALLY need / expect?
```
Answer: One or more URLs. It depends on where the data is available.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -13,58 +11,125 @@
+            metadata_dict = data_bundle['user_metadata']
+            data_objects = bundle['data_objects']
+        except KeyError:
+            logger.exception('Failed to parse bundle')
```
Is the exception message logged where the exception is caught and handled?
If so, as I expect it should be, then the message would not also need to be logged here as well.

jessebrennan replied:
Good catch! It looks like there was duplicate logging.

jessebrennan replied:
Actually, it's not quite duplicate since it's logging a different exception. Not sure whether or not to delete.

jessebrennan replied:
On more experimentation, just copying the message from the first exception won't tell you much about what actually went wrong parsing. Instead I'm using the `raise ... from` pattern so it's clear what happened. LMK if you disagree.

mikebaumann reviewed (changes requested):
Hi Jesse! Great work, again!
Here are a few comments to get started with as I continue the review.

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -74,65 +139,72 @@ def _load_bundle(self, bundle_uuid, metadata_dict, data_objects):
+                parsed_bundle = self._parse_bundle(bundle)
+                self.bundles_parsed.append(parsed_bundle)
+            except ParseError:
+                logger.exception(f'Could not parse bundle {count + 1}')
```
Because you know it is a `ParseError` and the `ParseErrors` have messages, I would log the message here, instead of explicitly logging the error before throwing the `ParseError`.

mikebaumann replied:
This comment has not been addressed yet.

jessebrennan replied:
Sorry I should have responded before. This functionality is actually already there. See https://docs.python.org/3/library/logging.html#logging.exception

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -74,65 +139,72 @@ def _load_bundle(self, bundle_uuid, metadata_dict, data_objects):
-                logger.info(f'Successfully loaded {len(bundles_loaded)} bundles')
+                logger.info(f'Successfully loaded {len(self.bundles_loaded)} bundles')
             else:
                 logger.info('Successfully loaded all bundles!')
```
Based on the logic above, it seems 'Successfully loaded all bundles!' would be logged even if there were unattempted bundles or bundles that could not be parsed, which seems inconsistent / misleading.

jessebrennan replied:
Perhaps the code has changed since you wrote this, but looking now I don't think it's true.

mikebaumann commented on tests/test_standard_loader.py:
```diff
@@ -0,0 +1,278 @@
+        cls.dss_client = hca.dss.DSSClient()
+        cls.dss_client.host = 'https://hca-dss-4.ucsc-cgp-dev.org/v1'
+        cls.dss_endpoint = os.getenv("TEST_DSS_ENDPOINT", "https://hca-dss-4.ucsc-cgp-dev.org/v1")
+        cls.staging_bucket = os.getenv("DSS_S3_STAGING_BUCKET", "mbaumann-dss-staging")
```
Please try changing `mbaumann-dss-staging` to `commons-dss-upload`.
This should work, and if not, please let me know and I'll review/adjust the IAM settings.

mikebaumann replied:
Please change:
```
cls.dss_client.host = 'https://hca-dss-4.ucsc-cgp-dev.org/v1'
cls.dss_endpoint = os.getenv("TEST_DSS_ENDPOINT", "https://hca-dss-4.ucsc-cgp-dev.org/v1")
```
to:
```
cls.dss_endpoint = os.getenv("TEST_DSS_ENDPOINT", "https://hca-dss-4.ucsc-cgp-dev.org/v1")
cls.dss_client.host = cls.dss_endpoint

```

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -13,58 +11,125 @@
 
 class StandardFormatBundleUploader:
     _uuid_regex = re.compile('[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}')
+    # adapted from http://mattallan.org/posts/rfc3339-date-time-validation/
```
Thanks for providing the reference!

mikebaumann commented on tests/test_standard_loader.py:
```diff
@@ -0,0 +1,278 @@
+                'data_bundle': {
+                    'id': bundle_uuid,
+                    'data_object_ids': file_guid,
+                    'created': tz_utc_now(),  # FIXME is this necessary?????
```
I think `created` is needed. What is `this` referring to?

jessebrennan replied:
Just left over development cruft. Thanks for catching it

mikebaumann reviewed (changes requested):
Again, this looks great, Jesse!
My main remaining concern is the oft-called method `_test_gen3_loading` will pass even if no data files or metadata.json were actually loaded. I realize this test was inherited (not written by you), but it seems like an important item to address to round-out the tests. I expect questions from you on this.

jessebrennan committed:
Make exception logging more clear

jessebrennan committed:
Implement review suggestions

jessebrennan committed:
Add more review suggestions

Summary of what's new:
- add test to make sure all objects in a bundle are loaded properly
- make loader tests inherit to remove duplication
- if any tests fail, false is returned and exit status will be zero
- renamed host bucket :)

jessebrennan commented:
@mikebaumann ready for a re-review!

jessebrennan commented:
I just rebased onto develop

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -198,11 +198,14 @@ def load_all_bundles(self, input_json: typing.List[dict]):
                 logger.error(f'Could not load {len(self.bundles_failed_parsed)} bundles')
+                success = False
                 # TODO: ADD COMMAND LINE OPTION TO SAVE ERROR LOG TO FILE https://stackoverflow.com/a/11233293/7830612
                 logger.info(f'Successfully loaded {len(self.bundles_loaded)} bundles')
```
This line should be removed from here and moved into the `if success` block, replacing the line that is currently there.

jessebrennan replied:
I agree it should be moved, but not there. The intention of this line was to tell the user exactly how many bundles were *actually* loaded, but only when it's different from the total number of bundles. It belongs in an else statement for success.

mikebaumann reviewed (changes requested):
Thanks, Jesse!
Here are a couple comments to get started on as I continue the review ...

mikebaumann commented on scripts/cgp_data_loader.py:
```diff
@@ -65,13 +65,17 @@ def main(argv=sys.argv[1:]):
 if __name__ == '__main__':
-    main()
+    success = main()
+    if not success:
```
Nice - thanks!

jessebrennan replied:
:+1:

mikebaumann commented on loader/standard_loader.py:
```diff
@@ -74,65 +136,77 @@ def _load_bundle(self, bundle_uuid, metadata_dict, data_objects):
                                                                  file_guid,
                                                                  file_version=file_version)
+            if already_present:
+                logger.info('File already present. No upload necessary.')
```
I think this could be logged at `debug` to be consistent with the related log messages around it.
Also, I think changing to the following would be appropriate:
```
    if already_present:
             logger.debug(f'...File already present, no upload necessary:  {filename} with uuid.version {file_uuid}.{file_version}')
    else
             logger.debug(f'...Successfully uploaded data file: {filename} with uuid.version {file_uuid}.{file_version}')
```

jessebrennan replied:
agreed! :+1:

mikebaumann reviewed (changes requested):
More comments ... let's talk!

jessebrennan committed:
Add better logging and control flow for file loading

jessebrennan committed:
Make logging more consistent

jessebrennan commented:
@mikebaumann Ok! I think I got everything. The part to pay attention to is the section in base_loader.py where different action is taken depending on the response code.

jessebrennan committed:
Improve logging and add clarity to upload

mikebaumann reviewed (approved):
Great work, Jesse! Thanks!
